### PR TITLE
added apis

### PR DESCRIPTION
**Pull request by massand**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -135,11 +135,206 @@ Each component talks to the other components over RPC of some sort (TODO: HTTP-o
...
+  }
+}
+```
+
+### Trigger IoT device reprovisioning flow
+`GET /identities/aziot/device/reprovision`
+
+#### Response
+```
+200 Ok
```
Comment (varunpuranik) [inline README.md:177]: Wondering about this API - how will the caller know reprovisioning is complete or failed, etc?

File: README.md [hunk 2]
```diff
@@ -135,11 +135,206 @@ Each component talks to the other components over RPC of some sort (TODO: HTTP-o
...
+```json
+{
+  "keystorehandle": "string"
+}
+```
+
+### Get Key
+`GET /keystore/key`
+
+#### Request
```
Comment (varunpuranik) [inline README.md:202]: Is this correct? Isn't this a Get?
Also do we need Get/Create?

File: README.md [hunk 3]
```diff
@@ -135,11 +135,206 @@ Each component talks to the other components over RPC of some sort (TODO: HTTP-o
...
+
+#### Response
+```json
+{
+  "keystorehandle": "string"
+}
+```
+
+### Create Key Pair
+`POST /keystore/keypair`
```
Comment (varunpuranik) [inline README.md:217]: This is public private key pair? How do I get the public key?

File: README.md (hunk 1 above)
Comment (massand) [reply README.md:177]: The caller is expected to call `GET /identities/aziot/device` until a 200 is received. This is also the approach to detect provisioning complete on first boot

File: README.md (hunk 2 above)
Comment (massand) [reply README.md:202]: Yes - this is a generic keystore API that accepts a `keyid` and returns a `keyhandle`.

File: README.md (hunk 3 above)
Comment (arsing) [reply README.md:217]: See the existing diagrams like https://github.com/arsing/edgev2/blob/ca513fc8cf89a95a0d1b9c3d8467f05c28f62f95/operation-mr-deviceca.png - `openssl-engine-ks` takes the key handle and gives you the `openssl::EVP_PKEY` structures corresponding to both.

Code example: https://github.com/arsing/libiothsm2/blob/ks/iotedged/src/main.rs#L69-L76

Comment (massand) [reply README.md:217]: I was trying to look for the same in the keygen APIs. But, I think the OpenSSL APIs enable you to do so. @arsing can confirm.

File: README.md [hunk 4]
```diff
@@ -135,11 +135,206 @@ Each component talks to the other components over RPC of some sort (TODO: HTTP-o
...
+```
+
+### Import Cert
+`GET /certstore/cert`
+
+#### Request
+```json
+{
+  "certid": "string",
+  "pem": string
```
Comment (massand) [inline README.md:332]: I'll fix this syntax.

File: README.md (hunk 2 above)
Comment (massand) [reply README.md:202]: > Also do we need Get/Create?

I haven't seen an example of Get-Redirect-Post yet in our APIs. So, I kept those two as separate calls.

File: README.md [hunk 5]
```diff
@@ -135,11 +135,192 @@ Each component talks to the other components over RPC of some sort (TODO: HTTP-o
...
+
+#### Response
+```
+200 Ok
+```
 
 ## KS
 
+### Create Key
+`POST /keystore/key`
```
Comment (arsing) [inline README.md:183]: Do we need the `/keystore` / `/certstore` prefixes when all URIs are going to be under them?

File: README.md [hunk 6]
```diff
@@ -135,11 +135,192 @@ Each component talks to the other components over RPC of some sort (TODO: HTTP-o
...
+`GET /identities/aziot/device`
+
+#### Response (SAS case)
+```json
+{
+  "id": "aziot://myhub.net/device/device01",
+  "managedBy": "aziot://myhub.net/",
+  "auth": {
+    "type": "sas",
+    "keystorehandle": "string",
```
Comment (arsing) [inline README.md:156]: These should also be camel-case for consistency.

File: operation-mr-deviceca.plantuml [hunk 7]
```diff
@@ -51,51 +51,60 @@ group Workload CA
...
 		"openssl-engine-ks" -> KS ++: sign("key handle", ...)
 		return OK(signature)
-		return OK(signed X509)
+		return OK(signed X509)	
 		return OK(privkey, server cert)
 		return OK(privkey, server cert)
 	end
 end
 
+group Host_Module API calls
```
Comment (arsing) [inline operation-mr-deviceca.plantuml:98]: This is just for "sign (HMAC-SHA256) a digest", right? Do you also want to add the "encrypt a secret" and "wants a server cert" flows?

File: README.md [hunk 8]
```diff
@@ -135,11 +135,192 @@ Each component talks to the other components over RPC of some sort (TODO: HTTP-o
...
+
 ## CS
 
-## MR
+### Create Cert
+`POST /certstore/cert`
+
+#### Request
+```json
+{
```
Comment (arsing) [inline README.md:265]: This should return the PEM in base64.

Comment (arsing) [inline README.md:265]: - If this takes a certid, it should be a PUT and the certid should be in the URI. If it's a POST then it shouldn't take a certid anywhere and auto-generate one server-side, and return that.

- It needs to take a CSR in base64 in the body.

- It doesn't need subjectname, publickeyhandle, effectivedate, expirationdate.

File: README.md [hunk 9]
```diff
@@ -135,11 +135,192 @@ Each component talks to the other components over RPC of some sort (TODO: HTTP-o
...
+
+#### Response
+```json
+{
+  "pem": "string"
+}
+```
+
+### Get Cert List
+`GET /certstore/cert`
```
Comment (arsing) [inline README.md:293]: Why is a list operation needed?

If it is needed, it should return the same as "Load Cert", ie an array of PEM strings.

File: README.md [hunk 10]
```diff
@@ -135,11 +135,192 @@ Each component talks to the other components over RPC of some sort (TODO: HTTP-o
...
+        "publickeyhandle": "string",
+        "effectivedate": "ISO-8601 string",
+        "expirationdate": "ISO-8601 string"
+    }
+  ]
+}
+```
+
+### Import Cert
+`GET /certstore/cert`
```
Comment (arsing) [inline README.md:312]: POST.

Also it makes sense for this to be the same URI as "Create cert", and the server could differentiate between create and import based on whether the request body has a CSR or a PEM. If you agree, merge this with the "Create" definition above.

File: README.md [hunk 11]
```diff
@@ -135,11 +135,192 @@ Each component talks to the other components over RPC of some sort (TODO: HTTP-o
...
+```
+200 Ok
+```
 
 ## KS
 
+### Create Key
+`POST /keystore/key`
+
+#### Request
```
Comment (arsing) [inline README.md:185]: Same as CS, POST means ID is server-generated. PUT means ID is client-provided, but then put it in the URI.

File: README.md [hunk 12]
```diff
@@ -135,11 +135,192 @@ Each component talks to the other components over RPC of some sort (TODO: HTTP-o
...
+
+### Sign using Private Key
+`POST /keystore/sign`
+
+#### Request
+```json
+{
+  "keystorehandle": "string",
+  "algo": "HMACSHA256",
+  "data": "string"
```
Comment (arsing) [inline README.md:245]: `s/data/digest/` to clarify that digest operation needs to happen on the client.

File: README.md [hunk 13]
```diff
@@ -135,11 +135,192 @@ Each component talks to the other components over RPC of some sort (TODO: HTTP-o
...
+
+#### Response
+```json
+{
+  "keystorehandle": "string"
+}
+```
+
+
+### Sign using Private Key
```
Comment (arsing) [inline README.md:237]: This is not just for private keys. Signing is for both symmetric and asymmetric keys. (In the first place, HMAC-SHA256 is not an asymmetric key signing algorithm.)

File: README.md [hunk 14]
```diff
@@ -135,11 +135,192 @@ Each component talks to the other components over RPC of some sort (TODO: HTTP-o
...
+`POST /keystore/sign`
+
+#### Request
+```json
+{
+  "keystorehandle": "string",
+  "algo": "HMACSHA256",
```
Comment (arsing) [inline README.md:244]: Needs to take parameters, so it's not just a string.

File: README.md [hunk 15]
```diff
@@ -22,7 +22,7 @@ IoT Edge v2 is made up of several independent services instead of one monolithic
 
 Each component talks to the other components over RPC of some sort (TODO: HTTP-over-UDS? gRPC-over-UDS? D-Bus? Something else?)
 
-
+The configuration / settings below are represented in YAML, but they will follow the settings file format for IS (i.e. TOML).
```
Comment (arsing) [inline README.md:25]: Just change them to TOML then :)

File: README.md (hunk 5 above)
Comment (massand) [reply README.md:183]: I saw them as separate services. Is your suggestion to merge them under one root namespace?

File: operation-mr-deviceca.plantuml (hunk 7 above)
Comment (massand) [reply operation-mr-deviceca.plantuml:98]: We do not need to support that for host modules at the moment. Parity is needed only for Edge modules.

File: README.md (hunk 5 above)
Comment (arsing) [reply README.md:183]: KS and CS and IS are separate server endpoints, right? So it's unambiguous which one is receiving the request, and you don't need the prefix to differentiate.

File: operation-mr-deviceca.plantuml (hunk 7 above)
Comment (arsing) [reply operation-mr-deviceca.plantuml:98]: Okay, but still put it in an inner `group Host module wants to sign (HMAC-SHA256) a digest` for clarity.

File: README.md (hunk 13 above)
Comment (arsing) [reply README.md:237]: Actually, the only operation we have for symmetric keys is encrypt, so let's keep it as "Sign using private key" for now. It's also a very different operation for symmetric keys vs asymmetric keys, so combining them into a single operation probably isn't the best idea anyway. Asymmetric keys take digest and produce signature. Symmetric keys take data and produce digest, which should not be called "signing".

But the other comments about this still apply. Specifically, HMAC-SHA256 is not an asymmetric key algorithm, and for the asymmetric key algorithms some of them need parameters, so the algorithm cannot be just a string. See https://github.com/arsing/libiothsm2/blob/ks/ksd/src/lib.rs#L78-L84

File: README.md (hunk 9 above)
Comment (massand) [reply README.md:293]: I am wondering if we have a need for cleanup at some point. This is definitely forward-looking and I think we should add one for key and keypair resources as well.

File: README.md (hunk 12 above)
Comment (massand) [reply README.md]: I modified it to digest

File: README.md (hunk 10 above)
Comment (massand) [reply README.md]: Merged. Good call.

**Pull request closed without merging**